File: src/lib/ocrConfig.ts
import AsyncStorage from "@react-native-async-storage/async-storage";
import { Platform } from "react-native";

import { config } from "./config";

const OCR_CONFIG_KEY = "alice_ocr_provider_config";

/** Web builds never embed a shared OCR key — users must bring their own. */
export function requiresCustomOcrConfig(): boolean {
  return Platform.OS === "web";
}

export interface OcrProviderConfig {
  baseUrl: string;
  apiKey: string;
  model: string;
}

export interface OcrProviderPreset {
  id: string;
  label: string;
  baseUrl: string;
  /** Suggested vision-capable model for the preset. */
  model: string;
  /** Short hint shown under the preset, e.g. where to get a key. */
  hint?: string;
}

/**
 * Mainstream OpenAI-compatible vision providers. All of them expose a
 * `/chat/completions` endpoint that accepts `image_url` content parts, so the
 * same request body works across them — only the base URL, key and model name
 * differ.
 */
export const OCR_PROVIDER_PRESETS: OcrProviderPreset[] = [
  {
    id: "zhipu",
    label: "智谱 GLM",
    baseUrl: "https://open.bigmodel.cn/api/paas/v4",
    model: "glm-4v-flash",
    hint: "open.bigmodel.cn",
  },
  {
    id: "openai",
    label: "OpenAI",
    baseUrl: "https://api.openai.com/v1",
    model: "gpt-4o-mini",
    hint: "platform.openai.com",
  },
  {
    id: "qwen",
    label: "通义千问 VL",
    baseUrl: "https://dashscope.aliyuncs.com/compatible-mode/v1",
    model: "qwen-vl-plus",
    hint: "dashscope.aliyuncs.com",
  },
  {
    id: "moonshot",
    label: "Kimi (Moonshot)",
    baseUrl: "https://api.moonshot.cn/v1",
    model: "moonshot-v1-8k-vision-preview",
    hint: "platform.moonshot.cn",
  },
  {
    id: "siliconflow",
    label: "硅基流动",
    baseUrl: "https://api.siliconflow.cn/v1",
    model: "Qwen/Qwen2-VL-7B-Instruct",
    hint: "siliconflow.cn",
  },
  {
    id: "openrouter",
    label: "OpenRouter",
    baseUrl: "https://openrouter.ai/api/v1",
    model: "google/gemini-flash-1.5",
    hint: "openrouter.ai",
  },
  {
    id: "ollama",
    label: "Ollama (本地)",
    baseUrl: "http://localhost:11434/v1",
    model: "llama3.2-vision",
    hint: "无需 KEY，需开启本地服务",
  },
  {
    id: "custom",
    label: "自定义",
    baseUrl: "",
    model: "",
  },
];

/** A config is usable only when all three fields are non-empty. */
export function isCustomOcrConfigSet(
  cfg: OcrProviderConfig | null | undefined,
): boolean {
  return Boolean(cfg && cfg.baseUrl.trim() && cfg.apiKey.trim() && cfg.model.trim());
}

let _cached: OcrProviderConfig | null = null;
let _loaded = false;

export async function loadOcrProviderConfig(): Promise<OcrProviderConfig | null> {
  try {
    const data = await AsyncStorage.getItem(OCR_CONFIG_KEY);
    if (!data) {
      _cached = null;
      _loaded = true;
      return null;
    }
    const parsed = JSON.parse(data) as Partial<OcrProviderConfig>;
    if (
      typeof parsed.baseUrl === "string" &&
      typeof parsed.apiKey === "string" &&
      typeof parsed.model === "string"
    ) {
      _cached = {
        baseUrl: parsed.baseUrl,
        apiKey: parsed.apiKey,
        model: parsed.model,
      };
    } else {
      _cached = null;
    }
    _loaded = true;
    return _cached;
  } catch {
    _cached = null;
    _loaded = true;
    return null;
  }
}

/** Sync accessor for the cached value; null until loadOcrProviderConfig() has run. */
export function getCachedOcrProviderConfig(): OcrProviderConfig | null {
  return _loaded ? _cached : null;
}

export async function saveOcrProviderConfig(
  cfg: OcrProviderConfig | null,
): Promise<void> {
  _cached = cfg;
  _loaded = true;
  try {
    if (cfg) {
      await AsyncStorage.setItem(OCR_CONFIG_KEY, JSON.stringify(cfg));
    } else {
      await AsyncStorage.removeItem(OCR_CONFIG_KEY);
    }
  } catch {
    // ignore — config is also kept in-memory for the running session
  }
}

/**
 * Resolve the effective OCR config: a complete custom config wins; otherwise
 * fall back to the bundled Zhipu config on native only. Web never uses a
 * built-in key (it would be public in the JS bundle).
 */
export function resolveOcrConfig(
  custom: OcrProviderConfig | null | undefined,
): { baseUrl: string; apiKey: string; model: string } | null {
  if (isCustomOcrConfigSet(custom)) {
    return {
      baseUrl: custom!.baseUrl.trim(),
      apiKey: custom!.apiKey.trim(),
      model: custom!.model.trim(),
    };
  }
  if (requiresCustomOcrConfig()) return null;
  if (!config.zhipuApiKey.trim()) return null;
  return {
    baseUrl: config.zhipuBaseUrl,
    apiKey: config.zhipuApiKey,
    model: config.visionModel,
  };
}

/** Whether OCR can run with the given custom config (or built-in on native). */
export function hasUsableOcrConfig(
  custom: OcrProviderConfig | null | undefined,
): boolean {
  return resolveOcrConfig(custom) !== null;
}

/** Whether the effective config is the user's custom one (not the built-in). */
export function isUsingCustomConfig(
  custom: OcrProviderConfig | null | undefined,
): boolean {
  return isCustomOcrConfigSet(custom);
}

/**
 * Build the `/chat/completions` URL from a base URL. Tolerates base URLs that
 * already end with the path so users can paste a full endpoint if they like.
 */
export function buildChatCompletionsUrl(baseUrl: string): string {
  const trimmed = baseUrl.trim().replace(/\/+$/, "");
  if (/\/chat\/completions$/.test(trimmed)) return trimmed;
  return `${trimmed}/chat/completions`;
}

File: lib/services/ocr_config.dart
import 'dart:convert';

import 'package:flutter/foundation.dart' show kIsWeb;

import 'prefs.dart';

/// 自定义 OCR 服务商配置（BYOK）。对应 RN 版 src/lib/ocrConfig.ts。
const String _ocrConfigKey = 'alice_ocr_provider_config';

/// Web 构建永不内嵌共享 OCR 密钥 —— 用户必须自备。
bool requiresCustomOcrConfig() => kIsWeb;

class OcrProviderConfig {
  const OcrProviderConfig({
    required this.baseUrl,
    required this.apiKey,
    required this.model,
  });

  final String baseUrl;
  final String apiKey;
  final String model;

  Map<String, dynamic> toJson() => {
        'baseUrl': baseUrl,
        'apiKey': apiKey,
        'model': model,
      };

  static OcrProviderConfig? fromJson(Object? raw) {
    if (raw is! Map) return null;
    final baseUrl = raw['baseUrl'];
    final apiKey = raw['apiKey'];
    final model = raw['model'];
    if (baseUrl is! String || apiKey is! String || model is! String) {
      return null;
    }
    return OcrProviderConfig(baseUrl: baseUrl, apiKey: apiKey, model: model);
  }
}

class OcrProviderPreset {
  const OcrProviderPreset({
    required this.id,
    required this.label,
    required this.baseUrl,
    required this.model,
    this.hint,
  });

  final String id;
  final String label;
  final String baseUrl;

  /// 该预设推荐的视觉模型。
  final String model;

  /// 简短提示，比如去哪儿拿 key。
  final String? hint;
}

/// 主流 OpenAI 兼容视觉服务商。它们都暴露接受 `image_url` 内容块的
/// `/chat/completions`，所以同一份请求体在各家之间通用 —— 只有
/// base URL、key 和模型名不同。
const List<OcrProviderPreset> kOcrProviderPresets = [
  OcrProviderPreset(
    id: 'zhipu',
    label: '智谱 GLM',
    baseUrl: 'https://open.bigmodel.cn/api/paas/v4',
    model: 'glm-4v-flash',
    hint: 'open.bigmodel.cn',
  ),
  OcrProviderPreset(
    id: 'openai',
    label: 'OpenAI',
    baseUrl: 'https://api.openai.com/v1',
    model: 'gpt-4o-mini',
    hint: 'platform.openai.com',
  ),
  OcrProviderPreset(
    id: 'qwen',
    label: '通义千问 VL',
    baseUrl: 'https://dashscope.aliyuncs.com/compatible-mode/v1',
    model: 'qwen-vl-plus',
    hint: 'dashscope.aliyuncs.com',
  ),
  OcrProviderPreset(
    id: 'moonshot',
    label: 'Kimi (Moonshot)',
    baseUrl: 'https://api.moonshot.cn/v1',
    model: 'moonshot-v1-8k-vision-preview',
    hint: 'platform.moonshot.cn',
  ),
  OcrProviderPreset(
    id: 'siliconflow',
    label: '硅基流动',
    baseUrl: 'https://api.siliconflow.cn/v1',
    model: 'Qwen/Qwen2-VL-7B-Instruct',
    hint: 'siliconflow.cn',
  ),
  OcrProviderPreset(
    id: 'openrouter',
    label: 'OpenRouter',
    baseUrl: 'https://openrouter.ai/api/v1',
    model: 'google/gemini-flash-1.5',
    hint: 'openrouter.ai',
  ),
  OcrProviderPreset(
    id: 'ollama',
    label: 'Ollama (本地)',
    baseUrl: 'http://localhost:11434/v1',
    model: 'llama3.2-vision',
    hint: '无需 KEY，需开启本地服务',
  ),
  OcrProviderPreset(
    id: 'custom',
    label: '自定义',
    baseUrl: '',
    model: '',
  ),
];

/// 三个字段都非空时配置才可用。
bool isCustomOcrConfigSet(OcrProviderConfig? cfg) {
  return cfg != null &&
      cfg.baseUrl.trim().isNotEmpty &&
      cfg.apiKey.trim().isNotEmpty &&
      cfg.model.trim().isNotEmpty;
}

OcrProviderConfig? _cached;
bool _loaded = false;

Future<OcrProviderConfig?> loadOcrProviderConfig() async {
  try {
    final data = await Prefs.getString(_ocrConfigKey);
    if (data == null || data.isEmpty) {
      _cached = null;
      _loaded = true;
      return null;
    }
    _cached = OcrProviderConfig.fromJson(json.decode(data));
    _loaded = true;
    return _cached;
  } catch (_) {
    _cached = null;
    _loaded = true;
    return null;
  }
}

/// 同步读取缓存值；loadOcrProviderConfig() 跑过之前为 null。
OcrProviderConfig? getCachedOcrProviderConfig() => _loaded ? _cached : null;

Future<void> saveOcrProviderConfig(OcrProviderConfig? cfg) async {
  _cached = cfg;
  _loaded = true;
  try {
    if (cfg != null) {
      await Prefs.setString(_ocrConfigKey, json.encode(cfg.toJson()));
    } else {
      await Prefs.remove(_ocrConfigKey);
    }
  } catch (_) {
    // 忽略 —— 内存里仍然是对的
  }
}

/// 从 base URL 拼出 `/chat/completions`。
/// 容忍已经带上该路径的写法，用户可以直接粘贴完整端点。
String buildChatCompletionsUrl(String baseUrl) {
  final trimmed = baseUrl.trim().replaceAll(RegExp(r'/+$'), '');
  if (RegExp(r'/chat/completions$').hasMatch(trimmed)) return trimmed;
  return '$trimmed/chat/completions';
}
